src/Elastic.Apm/Api/Destination.cs

```diff
@@ -0,0 +1,42 @@
+using Elastic.Apm.Report.Serialization;
+using Newtonsoft.Json;
+
+namespace Elastic.Apm.Api
+{
+	public class Destination
+	{
+		private Optional<string> _address;
+		private Optional<int?> _port;
+
+		[JsonConverter(typeof(TrimmedStringJsonConverter))]
+		public string Address
+		{
+			get => _address.Value;
+			set => _address = new Optional<string>(value);
+		}
+
+		public int? Port
+		{
+			get => _port.Value;
+			set => _port = new Optional<int?>(value);
+		}
+
+		internal void CopyMissingPropertiesFrom(Destination src)
+		{
+			if (!_address.HasValue) _address = src._address;
+			if (!_port.HasValue) _port = src._port;
+		}
+
+		private readonly struct Optional<T>
```

Comment:
Without `Optional` this test will not pass https://github.com/elastic/apm-agent-dotnet/pull/655/files#diff-3f1be8557a2341f7a6ee4692c1d9cd54R767

The goal is to allow public API user to prohibit automatic deduction of any of  `context.destination` properties. To achieve that we need a way to distinguish between `null` as the initial value (meaning public API user is okay with us automatically deducing it) and `null` explicitly set via public API.

I'll add above explanation as a comment.

Reply:
I see. Is there any specific reason we do it this way?

To be honest I'd keep it simple and to me this feels that we overcomplicate it. I'm not sure users will even realize that they can explicitly set these to `null` and we don't deduce these values in that case.

Put probably I won't be able to convince you and it's definitely not worth blocking this, so fine by me if you feel this is worth the complications. 

Maybe 1 thing I'd do: The comment on `Optional` is useful, so that makes me understand this. But that one is not public, so if you want users to know about this, I'd either comment something about this on the public `Destination` class or on the 2 public fields - given they are public it'd be useful to add XMLDocs to those anyway.

Reply:
I've added docs to `Destination`.

Reply:
> I've added docs to Destination. 

👍  

Wanna share some thoughts on why we do this? I'm really curious why we go into this at all.

Reply:
I'm not saying we need it 100% - I just want to make it a little bit future proof. For example the way Go agent implements it (by doing deduction eagerly when `Context.Http` or `Context.Db` is set/changed) allows user to nullify any deduced properties. I'm not saying users will definitely use it but I cannot say for sure they won't. And for our API to implement it the Go agent way will require to draw some kind connection between `SpanContext.Http`/`SpanContext.Db` and `SpanContext.Destination` will involve much more complication.
If you strongly feel that we should remove it I am okay with removing it.

Reply:
> If you strongly feel that we should remove it I am okay with removing it.

I'm ok with keeping - just wanted to know why we do this. Thanks for the answer.